StreamChatClient_v3/ChatClient.swift

```diff
@@ -222,8 +222,34 @@ public class Client<ExtraData: ExtraDataTypes> {
         }
     }
     
+    /// Sets a new **guest** user of the `ChatClient` as a current user.
+    ///
+    /// Guest sessions do not require any server-side authentication.
+    /// Guest users have a limited set of permissions.
+    ///
+    /// - Parameters:
+    ///   - userId: The new guest-user identifier.
+    ///   - extraData: The extra data of the new guest-user.
+    ///   - completion: The completion. Will be called when the new guest user is set. If setting up the new user fails the completion
+    /// will be called with an error.
     public func setGuestUser(userId: UserId, extraData: ExtraData.User? = nil, completion: ((Error?) -> Void)? = nil) {
-        fatalError("not implemented")
+        disconnect()
+        prepareEnvironmentForNewUser(userId: userId, role: .guest, extraData: extraData) { error in
+            guard error == nil else {
+                completion?(error)
+                return
+            }
+
+            self.apiClient.request(endpoint: .token(userId: userId, extraData: extraData)) {
+                switch $0 {
+                case let .success(payload):
+                    self.currentToken = payload.token
+                    self.connect(completion: completion)
+                case let .failure(error):
+                    completion?(error)
+                }
+            }
```

Comment:
What a nice retain cycle we have here :)

Reply:
Actually, there is no retain cycle here. String ref to `ChatClient` captured by a closure will be alive until the network request is finished as the lifetime of this closure is limited to the lifetime of the data-task. So it doesn't lead to a memory leak.

At the same time, you are Absolutely right that it doesn't make sense to keep `ChatClient` alive if the calling code doesn't have any/ drop all strong references to this object 👍

Reply:
Sorry, bad wording from my side. It's not a retain cycle. What I meant is the situation you described. It's not a big deal really, but it seems to me it's safer when the completion is not called, once the client is released.

Reply:
fixed in c2f9b54ca1e776672e3feafe43f9cca993534e08